File: backend/app/schemas/analytics.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from app.analytics.models import AnalyticsSummary, Anomaly, Metric, TrendResult


class AnalyticsRunRequest(BaseModel):
    dataset_id: str = Field(..., description="Uploaded dataset identifier")
    profile: str | None = Field(default=None, description="Optional dataset profile override")


class AnalyticsRunResponse(BaseModel):
    success: bool = True
    dataset_id: str
    profile: str
    capabilities: dict[str, bool] = Field(default_factory=dict)
    summary: AnalyticsSummary | None = None
    metrics: list[Metric] = Field(default_factory=list)
    trends: list[TrendResult] = Field(default_factory=list)
    anomalies: list[Anomaly] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
    raw: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/types/analytics.ts
export interface AnalyticsMetric {
  name: string
  value: number | null
  unit?: string | null
  description?: string
}

export interface AnalyticsTrendPoint {
  date: string
  value: number | null
}

export interface AnalyticsTrend {
  metric: string
  points: AnalyticsTrendPoint[]
  direction: string
}

export interface AnalyticsAnomaly {
  metric: string
  label: string
  value: number | null
  severity: string
  reason: string
}

export interface AnalyticsSummary {
  profile: string
  row_count: number
  column_count: number
  detected_capabilities: Record<string, boolean>
  metric_count: number
}

export interface BreakdownItem {
  label: string
  value: number
  percentage?: number
}

export interface ChartDataPoint {
  label: string
  value: number
}

export interface ReportDefinition {
  key: string
  title: string
  description: string
  profile: string
  required_capabilities: string[]
}

export interface ReportType extends ReportDefinition {
  available: boolean
  status: 'Available' | 'Unavailable'
}

export interface AnalyticsResponse {
  success: boolean
  dataset_id: string
  profile: string
  capabilities: Record<string, boolean>
  summary: AnalyticsSummary | null
  metrics: AnalyticsMetric[]
  trends: AnalyticsTrend[]
  anomalies: AnalyticsAnomaly[]
  notes: string[]
  raw?: Record<string, unknown>
}
